pinot-common/src/main/java/org/apache/pinot/common/utils/helix/IdealStateGroupCommit.java

```diff
@@ -0,0 +1,298 @@
+
+  /**
+   * Set up a group committer and its associated queues
+   */
+  public IdealStateGroupCommit() {
+    // Don't use Arrays.fill();
+    for (int i = 0; i < _queues.length; i++) {
+      _queues[i] = new Queue();
+    }
+  }
+
+  private Queue getQueue(String resourceName) {
+    return _queues[(resourceName.hashCode() & Integer.MAX_VALUE) % _queues.length];
+  }
+
+  public static synchronized void setMinNumCharsInISToTurnOnCompression(int minNumChars) {
+    _minNumCharsInISToTurnOnCompression = minNumChars;
+  }
+
+  /**
+   * Do a group update for idealState associated with a given resource key
+   * @param helixManager helixManager with the ability to pull from the current data\
+   * @param resourceName the resource name to be updated
+   * @param updater the idealState updater to be applied
+   * @return true if successful, false otherwise
+   */
+  public ObjectBooleanPair<IdealState> commit(HelixManager helixManager, String resourceName,
+      Function<IdealState, IdealState> updater, RetryPolicy retryPolicy, boolean noChangeOk) {
+    Queue queue = getQueue(resourceName);
+    Entry entry = new Entry(resourceName, updater);
+
+    boolean success = true;
+    queue._pending.add(entry);
+    IdealState finalIdealState = null;
+    while (!entry._sent.get()) {
+      if (queue._running.compareAndSet(null, Thread.currentThread())) {
+        ArrayList<Entry> processed = new ArrayList<>();
+        try {
+          if (queue._pending.peek() == null) {
+            return new ObjectBooleanImmutablePair<>(finalIdealState, true);
+          }
+          // remove from queue
+          Entry first = queue._pending.poll();
+          processed.add(first);
+          String mergedResourceName = first._resourceName;
+          HelixDataAccessor dataAccessor = helixManager.getHelixDataAccessor();
+          PropertyKey idealStateKey = dataAccessor.keyBuilder().idealStates(resourceName);
+          IdealState idealState = dataAccessor.getProperty(idealStateKey);
+
+          // Make a copy of the idealState above to pass it to the updater
+          // NOTE: new IdealState(idealState.getRecord()) does not work because it's shallow copy for map fields and
+          // list fields
+          IdealState idealStateCopy = HelixHelper.cloneIdealState(idealState);
+
+          /**
+           * If the local cache does not contain a value, need to check if there is a
+           * value in ZK; use it as initial value if exists
+           */
+          IdealState updatedIdealState = first._updater.apply(idealStateCopy);
+          Iterator<Entry> it = queue._pending.iterator();
+          while (it.hasNext()) {
+            Entry ent = it.next();
+            if (!ent._resourceName.equals(mergedResourceName)) {
+              continue;
+            }
+            processed.add(ent);
+            updatedIdealState = ent._updater.apply(idealStateCopy);
+            // System.out.println("After merging:" + merged);
+            it.remove();
```

Comment:
Is it safe to modify the queue here while we are iterating on it?

Reply:
yes, this is processed in single thread